lib/executor/variable_arrival_rate_test.go

```diff
@@ -0,0 +1,363 @@
+		StartRate: null.IntFrom(10),
+		Stages: []Stage{
+			{
+				Duration: types.NullDurationFrom(time.Second * 1),
+				Target:   null.IntFrom(10),
+			},
+			{
+				Duration: types.NullDurationFrom(time.Second * 1),
+				Target:   null.IntFrom(50),
+			},
+			{
+				Duration: types.NullDurationFrom(time.Second * 1),
+				Target:   null.IntFrom(50),
+			},
+		},
+		PreAllocatedVUs: null.IntFrom(10),
+		MaxVUs:          null.IntFrom(20),
+	}
+	logHook := &testutils.SimpleLogrusHook{HookedLevels: []logrus.Level{logrus.WarnLevel}}
+	testLog := logrus.New()
+	testLog.AddHook(logHook)
+	testLog.SetOutput(ioutil.Discard)
+	logEntry := logrus.NewEntry(testLog)
+	es := lib.NewExecutionState(lib.Options{}, 10, 50) //TODO: fix
+	runner := lib.MiniRunner{
+		Fn: vuFn,
+	}
+
+	es.SetInitVUFunc(func(_ context.Context, _ *logrus.Entry) (lib.VU, error) {
+		return &lib.MiniRunnerVU{R: runner}, nil
+	})
+
+	initializeVUs(ctx, t, logEntry, es, 10)
+
+	executor, err := config.NewExecutor(es, logEntry)
+	require.NoError(t, err)
+	err = executor.Init(ctx)
+	require.NoError(t, err)
+	return ctx, cancel, executor, logHook
+}
+
+func TestVariableArrivalRateRunNotEnoughAlloctedVUsWarn(t *testing.T) {
+	t.Parallel()
+	var ctx, cancel, executor, logHook = testVariableArrivalRateSetup(
+		t, func(ctx context.Context, out chan<- stats.SampleContainer) error {
+			time.Sleep(time.Second)
+			return nil
+		})
+	defer cancel()
+	var engineOut = make(chan stats.SampleContainer, 1000)
+	err := executor.Run(ctx, engineOut)
+	require.NoError(t, err)
+	entries := logHook.Drain()
+	require.NotEmpty(t, entries)
+	for _, entry := range entries {
+		require.Equal(t,
+			"Insufficient VUs, reached 20 active VUs and cannot allocate more",
+			entry.Message)
+		require.Equal(t, logrus.WarnLevel, entry.Level)
+	}
+}
+
+func TestVariableArrivalRateRunCorrectRate(t *testing.T) {
+	t.Parallel()
+	var count int64
+	var ctx, cancel, executor, logHook = testVariableArrivalRateSetup(
+		t, func(ctx context.Context, out chan<- stats.SampleContainer) error {
+			atomic.AddInt64(&count, 1)
+			return nil
+		})
+	defer cancel()
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// check that we got around the amount of VU iterations as we would expect
+		var currentCount int64
+
+		time.Sleep(time.Second)
+		currentCount = atomic.SwapInt64(&count, 0)
+		require.InDelta(t, 10, currentCount, 1)
+
+		time.Sleep(time.Second)
+		currentCount = atomic.SwapInt64(&count, 0)
+		// this is highly dependant on minIntervalBetweenRateAdjustments
+		// TODO find out why this isn't 30 and fix it
+		require.InDelta(t, 23, currentCount, 2)
+
+		time.Sleep(time.Second)
+		currentCount = atomic.SwapInt64(&count, 0)
+		require.InDelta(t, 50, currentCount, 2)
+	}()
+	var engineOut = make(chan stats.SampleContainer, 1000)
+	err := executor.Run(ctx, engineOut)
+	wg.Wait()
+	require.NoError(t, err)
+	require.Empty(t, logHook.Drain())
+}
+
+func TestVariableArrivalRateCancel(t *testing.T) {
```

Comment:
This test and `TestConstantArrivalRateCancel` are line-for-line identical save for the constant/variable differences.

Could we reuse a generic function for both to avoid the repetition?

Reply:
:+1: managed to do it without as much hassle as I expected. Will wait for the other comments to resolve so I only force-push ... once ...

Reply:
Hhmm do we want to make them part of the same test, though? Now a single test (`TestArrivalRateCancel`) is testing two executors, and it's weird since the variable rate executor is partially tested in `constant_arrival_rate_test.go`. I'd rather have them be separate tests that happen to share the same generic helper function. It might be a slight difference, but I think they should be separate tests.

Reply:
if we have it in two separate functions it will be something like:
```
TestConstantArrivalRateCancel(....){
 testArrivalRateCancel(t, testConstantArrivalRateSetup)
} 
```

Which to me seems even worse ... as now looking at this test you need to find both `testArrivalRateCancel` and `testConstantArrivalRateSetup` and the second one is already used in other places so I prefer if `testArrivalRateCancel` doesn't also gets moved around. If you prefer I could move this to a separate `common_arrival_rate_test.go` ?

Reply:
Hhmm well "find" is simply a matter of "Go to definition" in your IDE, so I woudn't mind that first approach you suggested. Though a `common_arrival_rate_test.go` would make sense.

Not sure, I don't feel strongly either way. If you think it's fine leaving it as is, I'd be OK with it if the alternatives are messier. Up to you :)

Reply:
I am going to leave it for now, maybe we will add similar tests for the other Executors and we will refactor it then ;)